src/language_features/document_symbol.rs

```diff
@@ -206,6 +207,24 @@ fn editor_document_symbol(
     ctx.exec(meta, command);
 }
 
+enum Tree {
+    Child,
+    LastChild,
+    Pipe,
+    Empty,
+}
+
+impl fmt::Display for Tree {
```

Comment:
> 1. Perhaps, previous style was there for a reason, e.g. to make it easy to parse the buffer in case if users would want to customize it using Kakoune itself.

Not specifically.

> And probably, it already could be implemented with highlighters. I'm still learning Kakoune, so not 100% sure about this.

Highlighters can work but are not the best choice (you can't navigate or search for the virtual text as created by replace-ranges).
Users can indeed override `lsp-show-document-symbol` but barely anyone does.

> 3. Use of UTF-8 symbols instead of ASCII.

That's fine because Kakoune already uses box drawing characters unconditionally for `:info`.

---

The reason why symbol kinds are in a separate column is because their length is effectively bounded (the longest one is "TypeParameter") so it is cheap -- in terms of screen real estate -- to align them.
I think with the tree, the alignment is no longer important enough, so I'd add the kind directly after the symbol name (or if we have universally-understood icons we can also put them before).

Also the new tree clearly outshines the indented approach in the first column. I don't think that one adds value anymore, especially because it's all the same file.
The correspondence between nesting structure and line number makes no sense. Only between nesting structure and symbol, or symbol and line number.

So how about

```
links-fixer.scala:17:12: Link (Class)
links-fixer.scala:24:8:  LinksFixer (Module)
links-fixer.scala:26:7:  ├── main (Method)
links-fixer.scala:29:15: ├── fixFile (Method)
links-fixer.scala:30:9:  │   └── file (Constant)
links-fixer.scala:34:15: ├── fix (Method)
links-fixer.scala:37:15: ├── parse (Method)
links-fixer.scala:38:9:  │   ├── trimmed (Constant)
links-fixer.scala:39:15: │   ├── prefix (Constant)
links-fixer.scala:39:23: │   ├── title (Constant)
links-fixer.scala:40:9:  │   ├── fragment (Constant)
links-fixer.scala:44:9:  │   ├── fragmentless (Constant)
links-fixer.scala:45:10: │   ├── path (Constant)
links-fixer.scala:45:16: │   └── file (Constant)
links-fixer.scala:51:15: └── render (Method)
```

WDYT?

(the file path is still noisy but it's a reasonable compromise to make this look and feel like a grep-buffer)

Reply:
Agree, pushed a commit.

Reply:
I think [Visual Studio Code icons](https://github.com/microsoft/vscode-icons/) are probably the most recognizable?

Although, they are in .svg format.

Another option would be to highlight the `SymbolKind` with the same color as in the editor, but syntax highlighting is out of the scope of the project. 

Or maybe just coloring them with the same built-in terminal color per-type?

Reply:
> I think Visual Studio Code icons are probably the most recognizable?
> 
> Although, they are in .svg format.

Interesting, yeah.

I personally don't find them very intuitive (a method is a cube?) but it's
probably hard to do better with a small icon, and perhaps people who used
graphical IDEs recognize the pattern.

> 
> Another option would be to highlight the SymbolKind with the same color as in the editor, but syntax highlighting is out of the scope of the project.
> 
> Or maybe just coloring them with the same built-in terminal color per-type?

If the language server supports semantic tokens, then we can definitely apply the mapped face to the symbol name.
Our default configuration includes

```toml
[semantic_tokens]
faces = [
    {face="function", token="function"},
    {face="module", token="namespace"},
    {face="type", token="type"},
]
```

which means that given

```rust
mod some_module {
	fn some_function() {
	}
	struct SomeType {}
}
```

`some_module` has LSP semantic token kind "namespace", so we would add a highlighter that gives it the "module" Kakoune face.
The Kakoune theme sets the face. These are the available faces

```kak
# For Code
face global value red
face global type yellow
face global variable green
face global module green
face global function cyan
face global string magenta
face global keyword blue
face global operator yellow
face global attribute green
face global comment cyan
face global documentation comment
face global meta magenta
face global builtin default+b
```

For every SymbolKind we'd have to find a Semantic Token type and a fitting face.
This sounds like something we can do.
Not sure whether to keep the `(Function)` annotations, I guess so (I usually don't use syntax highlighting).

---

Instead of

	src/text_edit.rs:16:11:  TextEditish (Interface)
	src/text_edit.rs:17:8:   ├── text_edit (Function)
	src/text_edit.rs:18:8:   ├── as_ref (Function)
	src/text_edit.rs:19:8:   └── from_text_edit (Function)
	src/text_edit.rs:492:8:  KakouneTextEdit (Struct)
	src/text_edit.rs:493:5:  ├── range (Field)
	src/text_edit.rs:494:5:  ├── new_text (Field)
	src/text_edit.rs:495:5:  └── command (Field)

for Rust, I'd expect:

	src/text_edit.rs:16:11:  trait TextEditish
	src/text_edit.rs:17:8:   ├── fn text_edit
	src/text_edit.rs:18:8:   ├── fn as_ref
	src/text_edit.rs:19:8:   └── fn from_text_edit
	src/text_edit.rs:492:8:  struct KakouneTextEdit
	src/text_edit.rs:493:5:  ├── range: Range
	src/text_edit.rs:494:5:  ├── new_text: String
	src/text_edit.rs:495:5:  └── command: KakouneTextEditCommand

but that'd require language-specific info, it would be better if the LSP server did that

Reply:
> Faces

Yes, that is a good idea. It will cause integration problems with non-standard highlighters like https://github.com/phaazon/kak-tree-sitter, but such issues are better to be handled on their side, I think.

> Language-dependent outline

That would be a really good UX/UI improvement! To show the basic signature:
`pub fn as_ref`, `private val a: String`, `override def toString: String`.

Unfortunately, it seems like an LSP limitation.

I came to Kakoune after Intellij IDEA and that's how it looks there:
![image](https://github.com/kak-lsp/kak-lsp/assets/12570166/09bbbb0f-4050-48cf-8163-8723817cd3eb)


VS Code:
![image](https://github.com/kak-lsp/kak-lsp/assets/12570166/ff08a808-d753-4211-9db6-61e9bf1a7aa1)


So, VS Code shows the same amount of information as we do, as it also uses LSP.

IntelliJ IDEA doesn't use LSP and instead has its own internal way of integrating with languages.